rclpy/src/rclpy_common/src/handle.c

```diff
@@ -0,0 +1,146 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Python.h>
+#include <stddef.h>
+
+#include "rcutils/allocator.h"
+#include "rcutils/strdup.h"
+#include "rcutils/types/rcutils_ret.h"
+
+#include "rclpy_common/handle.h"
+
+struct rclpy_handle_t
+{
+  void * ptr;  // opaque pointer to the wrapped object.
+  size_t ref_count;  // Reference count.
+  struct rclpy_handle_t ** dependencies;  // array of pointers to dependencies.
+  size_t num_of_dependencies;  // size of the array.
+  rclpy_handle_destructor_t destructor;  // destructor
+};
+
+/// Creates a Handle object.
+rclpy_handle_t *
+_rclpy_create_handle(void * ptr, rclpy_handle_destructor_t destructor)
+{
+  assert(ptr);
+  assert(destructor);
+
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  rclpy_handle_t * handle = allocator.zero_allocate(1, sizeof(rclpy_handle_t), allocator.state);
+  if (!handle) {
+    PyErr_Format(PyExc_MemoryError, "Failed to allocate memory for handle");
+    return NULL;
+  }
+
+  handle->ptr = ptr;
+  handle->ref_count++;
+  handle->destructor = destructor;
+
+  return handle;
+}
+
+/// Adds a dependency to a handle.
+/**
+ * The `dependency` handle reference count is incresead.
+ * The `dependent` handle stores `dependency` into its dependencies list.
```

Comment:
Do we want to check reference cycles?
I would rather not handle that case (like `std::shared_ptr`), and carefully check them when reviewing.

Reply:
I agree with @ivanpauno. Preventing direct reference cycles only gives a false sense of safety, and growing this to become another `gc` is unnecessary. We just have to be careful. I will say though that I don't see the need for this API to be exposed. It could be handled entirely within the C extension, where we can be careful.